### ISSUE TITLE
Save chapter name when highlighting text

### ISSUE THREAD

**Issue by Galunid** [koreader/koreader#6138]
Is it possible to add additional entry to <ebook_name>.sdr/metadata.epub.lua that also saves the chapter where highlight was made? I'm working on a wikia page for one book, and it would make my life a lot easier. Not sure if there are already functions that take xpath for example and translate it into ToC entry. I'd be grateful if someone could point me in the right direction.

**Comment by poire-z** [comment]
Recent related issue: #6074 .

crengine might be able to give you that, but I think you should use the TOC from readertoc, which might be a bit more cleaned up than the one crengine knows.
I guess you could add there a method, which, from the current page and your highlight start xpointer, could:
- get current page TOC and its xpointer as the starting point
- use `self.ui.document:compareXPointers()` (see its use in readerhighlight.lua) to compare them
- get previous or next ones and do the comparisons until you find one xpointer from TOC that is just <= to your highlight xpointer.

**Comment by Galunid** [comment]
Actually, I've found that `ReaderToc:getTocTitleByPage(pn_or_xp)` works pretty nicely. Is there a way to run a function on a document when opened for the first time after update, or do I have to implement it myself? What I want to do is more or less this:
```lua
function ReaderHighlight:regenerateHighlights()
  for _, highlights in pairs(self.view.highlight.saved) do
    for _, highlight in pairs(highlights) do
      local chapter_name = self.ui.toc:getTocTitleByPage(highlight.pos0)
      highlight.chapter = chapter_name -- add new property(required by my patch to Evernote plugin)
    end
  end
end
```

**Comment by poire-z** [comment]
> works pretty nicely

May be, but it won't be accurate: if a new chapter starts in the middle of the page, and you highlight some stuff in the upper part, it will tell your highlight is from the next chapter (while it's part of the previous one).

> Is there a way to run a function on a document when opened for the first time after update, or do I have to implement it myself?

(We have `SettingsMigration:migrateSettings()` in ReaderUI, where this could fit, but it's run too early (before the document is loaded) for it to help you.)
But the question is whether you want to impose this update of all their highlights to all users - while there is no use (yet?) for the new field in KOReader (so, may be this hack could be just for you if you need that for past documents).
Another option (that I prefer) would be to just update this new field if it does not exist when the user open the Bookmarks view, where you're already looping thru all bookmarks.

**Comment by pazos** [comment]
Also related: #4566

**Comment by Galunid** [comment]
> But the question is whether you want to impose this update of all their highlights to all users - while there is no use (yet?) for the new field in KOReader (so, may be this hack could be just for you if you need that for past documents).
I was thinking of imposing this update on all users. It shouldn't break anything, as it only adds an extra field, without removing previous ones. I have working `exportToHTML` function for Evernote, `exportToJSON` should work without changing anything. I already "hacked" the documents I needed, I just thought, that if I have that stuff working, I can as well add this as a feature. I can probably add "fixing" the old files, while iterating in the bookmarks menu, although I don't have the slightest idea as to how to document that.

**Comment by poire-z** [comment]
> I can as well add this as a feature. I can probably add "fixing" the old files, while iterating in the bookmarks menu, although I don't have the slightest idea as to how to document that

It does not need any documenation I think. It's a new feature, it's allowed to work only on new documents - and if you update old highlights on showing Bookmarks (and may be also when you need it: in that exportToHTML if that happens on the current book opened), it's just fine to have all this happening without any documentation.

**Comment by poire-z** [comment]
> I have working exportToHTML function for Evernote

You're going to have that fixed and working well? :) https://github.com/koreader/koreader/issues/1095#issuecomment-453726356

**Comment by Galunid** [comment]
Are you referring to the exporting "notes"? I believe it's already working(#5440).

**Comment by Galunid** [comment]
![image](https://user-images.githubusercontent.com/10298730/81553512-80428d00-9385-11ea-9441-f7961efb0664.png)
This is what I have so far. First rectangle, case when there is no chapter, second when there is. Not sure how to format this, but I'm open to ideas.

**Comment by poire-z** [comment]
Oh, ok, good then :) (I haven't tested it since).
And I just see there is a templating system to process note.tpl and build the HTML - so I guess you'll be adding there the chapter name and have some HTML fun.

**Comment by poire-z** [comment]
> Not sure how to format this, but I'm open to ideas.

HTML fun that is :)
I guess page + chapter name on the left, and date on the right would make more sense.

I had never played with this export function (except this one time) - but a thing that might be nice would be to be able to build and open that html document from the book you're currently reading (without having to know all that is provided by the evernote plugin?). Dunno if that's already possible.
Anyway, please check that the resulting HTML renders well with crengine :)

**Comment by Galunid** [comment]
![image](https://user-images.githubusercontent.com/10298730/81555761-699e3500-9389-11ea-9b04-3dc2250d65c6.png)
Other than the missing yellow rectangle(where the red one is), it renders correctly, I think that's a pretty cool idea actually. I'm thinking of changing the notification after exporting to a widget that allows yes/no choice. Another option is a menu entry for Evernote plugin. Is that what you have in mind?

**Comment by poire-z** [comment]
> Other than the missing yellow rectangle(where the red one is), 

Yep, crengine does not support colors specified as `rgb(r,g,b)`:
https://github.com/koreader/koreader/blob/6e9de06fabfee79b5881ca9f4100877fc928f019/plugins/evernote.koplugin/note.tpl#L41
which is what the include timecolor() function builds:
https://github.com/koreader/koreader/blob/6e9de06fabfee79b5881ca9f4100877fc928f019/plugins/evernote.koplugin/note.tpl#L2-L4
but it might be easy to change it to output colors as `#rrggbb` instead (although we won't see these colors on our B&W eInk :), but I guess it's good to have)

> I'm thinking of changing the notification after exporting to a widget that allows yes/no choice. Another option is a menu entry for Evernote plugin. Is that what you have in mind?

I really had nothing in mind :) but yes, both options looks fine. I think (if it works really well and we want more people to see this feature exists), we can even have the plugin export another menu item that could be plugged in the main reader top menu just under `Bookmarks` (via reader_menu_order.lua).

**Comment by Galunid** [comment]
```diff
 function ReaderBookmark:onShowBookmark()
+    -- check if highlights already have "chapter" property
+    local has_chapter_property = false
+    for _, highlights in pairs(self.view.highlight.saved) do
+        for _, highlight in ipairs(highlights) do
+            if highlight.chapter then
+              has_chapter_property = true
+            end
+            break
+        end
+        break
+    end
+ 
+    if not has_chapter_property then
+        self:regenerate_highlights()
+    end
```
So, basically that's what I came up with, idea is that it takes first element of `self.view.highlight.saved` and checks if it has `chapter` property. I was looking for a way to do that without 2 loops, but I couldn't come up with anything. This double loop executes only once, but it looks pretty terrible. Not sure if there is a way to get rid of that monster though.
As for changing from rgb to hex, it's already done. I never wrote a plugin for koreader, but I'll try to take a look.

**Comment by poire-z** [comment]
You could put all the stuff in another method `ReaderBookmark:updateBookmarksIfNeeded()` to keep this one cleaner without this noise.
And just store the fact that you have migrated/added chapters by storing the fact as a setting:
```lua
function ReaderBookmark:updateBookmarksIfNeeded()
    if G_reader_settings.has("bookmarks_chapters_updated") then
       return
    end
    -- do your ugly stuff :)
end
```
Or even a setting with a datestamp (so it could be used by #5794 if it resurrects):
```lua
local version = G_reader_settings.readSetting("bookmarks_update_version") or 20101111
if version < 20200511 then
   -- do chapter migration
end
G_reader_settings.saveSetting("bookmarks_update_version", 20200511)
```

**Comment by Galunid** [comment]
The problem with this approach is that it will regenerate those properties for only one book(as far as I know, those properties are shared across the books, please correct me if I'm wrong). One option would be to "rebuild" all the books on update, although it would probably take some time. Another option is what I posted earlier, although refactored into it's own function, and I realized I can skip second `for` loop, because I can just take first item of the `highlights` array/list. Last one I can think of would be to add a menu item for user to forcefully rebuild it for the books they need.

**Comment by poire-z** [comment]
Oups, sorry, I'm so so used to mostly using `G_reader_settings` :)
Indeed, you should use `self.ui.doc_settings` instead!
```lua
local version = self.ui.doc_settings.readSetting("bookmarks_update_version") 
...
```
